src/Bicep.Core/Semantics/Namespaces/SystemNamespaceType.cs

```diff
@@ -629,6 +705,43 @@ private static SyntaxBase CreateStringLiteral(TypeSymbol typeSymbol)
             return SyntaxFactory.CreateStringLiteral(stringLiteral.RawStringValue);
         }
 
+        private static SyntaxBase JsonContentFunctionReturnTypeEvaluator(FunctionCallSyntaxBase functionCall, Symbol symbol, TypeSymbol typeSymbol, FunctionVariable? functionVariable, object? functionValue)
+        {
+            if (functionVariable is null)
+            {
+                // TemplateEmitter when emitting ARM-json code instead function will use createObject functions instead putting raw JSON.
+                // This can be avoided using functionVariables where bicep syntax is processed by Emitter itself, not by ExpressionConverter.
+                throw new InvalidOperationException($"Function Variable must be used");
+            }
+
+            return SyntaxFactory.CreateExplicitVariableAccess(functionVariable.Name);
+
+        }
+
+        private static SyntaxBase JsonContentFunctionVariableGenerator(FunctionCallSyntaxBase functionCall, Symbol symbol, TypeSymbol typeSymbol, bool directVariableAssignment, object? functionValue)
+        {
+            //converting JSON to bicep syntax and then back to ARM-JSON escapes ARM template expressions (`[variables('')]`, etc.) out of the box
+            return ConvertJsonToBicepSyntax(functionValue as JToken ?? throw new InvalidOperationException($"Expecting function to return {nameof(JToken)}, but {functionValue?.GetType().ToString() ?? "null"} received."));
+        }
+
+        private static SyntaxBase ConvertJsonToBicepSyntax(JToken token) =>
```

Comment:
There are a few places where we're doing something similar - in the decompiler, and in the 'insert resource' implementation. Is it possible to commonize this logic in a shared method?

Reply:
Insert resource uses System.Text.Json, here we use newtonsoft - mainly as code for convering json to type was already here implemented, it has selectToken and with @majastrz we agreed, that since backend uses newtonsoft, bicep should do to. Plus Newtonsoft is less grumpy about invalid jsons :) 

Decompiler has much more logic than just reading JSON as object, so it's not quite handy here. If we would somehow load the ARM-JSON using loadJsonContent - we should not make it bicep - all ARM expressions are needed to be treated as strings.

But in general yes - functions used in insert resource and here can be this same.

Reply:
Good point, thanks for the explanation!